src/lib/storage/storage_manager.cpp

```diff
@@ -275,110 +439,93 @@ std::ostream& operator<<(std::ostream& stream, const StorageManager& storage_man
+  export_value(static_cast<uint32_t>(segment->attribute_vector()->size()), file_name);
 
-void export_compressed_vector(const CompressedVectorType type, const BaseCompressedVector& compressed_vector,
-                              std::string file_name) {
-  switch (type) {
-    case CompressedVectorType::FixedWidthInteger4Byte:
-      export_values(dynamic_cast<const FixedWidthIntegerVector<uint32_t>&>(compressed_vector).data(), file_name);
-      return;
-    case CompressedVectorType::FixedWidthInteger2Byte:
-      export_values(dynamic_cast<const FixedWidthIntegerVector<uint16_t>&>(compressed_vector).data(), file_name);
-      return;
-    case CompressedVectorType::FixedWidthInteger1Byte:
-      export_values(dynamic_cast<const FixedWidthIntegerVector<uint8_t>&>(compressed_vector).data(), file_name);
-      return;
-    case CompressedVectorType::BitPacking:
-      export_compact_vector(dynamic_cast<const BitPackingVector&>(compressed_vector).data(), file_name);
-      return;
-    default:
-      Fail("Any other type should have been caught before.");
-  }
+  export_values(*segment->fixed_string_dictionary(), file_name);
+  export_compressed_vector(*segment->compressed_vector_type(), *segment->attribute_vector(), file_name);
+  //TODO: What to do with non-compressed AttributeVectors?
 }
 
-void StorageManager::write_dict_segment_to_disk(const std::shared_ptr<DictionarySegment<int>> segment,
-                                                const std::string& file_name) {
+template <typename T>
+void StorageManager::_write_dict_segment_to_disk(const std::shared_ptr<DictionarySegment<T>> segment,
+                                                 const std::string& file_name) const {
   /*
    * For a description of how dictionary segments look, see the following PR:
    *    https://github.com/hyrise-mp-22-23/hyrise/pull/94
    */
-  const auto compressed_vector_type_id = static_cast<uint32_t>(BinaryWriter::_compressed_vector_type_id<int>(*segment));
-  export_value(compressed_vector_type_id, file_name);
+  const auto compressed_vector_type_id =
+      _resolve_persisted_segment_encoding_type_from_compression_type(segment->compressed_vector_type().value());
+  export_value(static_cast<uint32_t>(compressed_vector_type_id), file_name);
   export_value(static_cast<uint32_t>(segment->dictionary()->size()), file_name);
   export_value(static_cast<uint32_t>(segment->attribute_vector()->size()), file_name);
 
-  export_values<int32_t>(*segment->dictionary(), file_name);
+  // we need to ensure that every part can be mapped with a uint32_t map
+  export_values<T>(*segment->dictionary(), file_name);
   export_compressed_vector(*segment->compressed_vector_type(), *segment->attribute_vector(), file_name);
+  //TODO: What to do with non-compressed AttributeVectors?
+}
+
+void StorageManager::_write_segment_to_disk(const std::shared_ptr<AbstractSegment> abstract_segment,
```

Comment:
This is worth a discussion of the responsibilities of different classes/objects.

Should the StorageManager handle the whole persisting process and know detailed what to do per SegmentClass?

Or would it be nicer to let each Segment implement a `.byte_representation()` method, which would return a byte array to write out to disk. The Segments already know how to initialize (decode) themselves from a byte array and therefore should be able to encode themselves. 
+ this would keep knowledge about the internal structure in the Segments
+ this would decouple implementation and allow for easy extension

In conclusion, I would strongly argue for the second option.

Reply:
100% agree.

Reply:
How do you like the way it is done now?

Reply:
I agree with the second option and would vote against having the encoding/decoding logic for each segment type in the storage manager.

Alternatively, one could develop a BinarySegmentEncoded/-Decoder that encapsulates the encoding/decoding logic.